File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-0.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        if (!head) return nullptr;
        Node* p2 = head;

        while(p2){
            Node* newN = new Node(p2->val);
            // newN->val = p2->val;
            newN->next = p2->next;
            p2->next = newN;
            p2 = newN->next;
        }
        p2 = head;
        Node* p1 = head->next; 
        while(p2){
            p1->random = p2->random ? p2->random->next : nullptr;
            p1 = p1->next? p1->next->next: nullptr;  
            p2 = p2->next? p2->next->next: nullptr;   
        }
        p1 = head->next;
        p2 = head;
        Node* ans = p1;
        while(p2){
            p2->next = p2->next->next;  
            p1->next = p1->next ? p1->next->next : nullptr;  
            p2 = p2->next;
            p1 = p1->next;
        }

        return ans; 
    }
};

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        self.ans = float('-inf')

        def ms(root):
            if not root:
                return 0
            
            ls = ms(root.left)
            rs = ms(root.right)

            temp = ls+rs+root.val
            self.ans = max(self.ans, temp)

            if (max(ls, rs) + root.val) > 0:
                return (max(ls, rs) + root.val)
            
            elif root.val > 0:
                return root.val
            
            else:
                return 0
        
        ms(root)
        return self.ans